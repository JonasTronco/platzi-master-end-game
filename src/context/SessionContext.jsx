import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
