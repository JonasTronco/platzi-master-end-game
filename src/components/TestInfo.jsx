import PropTypes from 'prop-types';
import axios from 'axios';
import Router from 'next/router';
import { useSession } from '../context/SessionContext';
import useRequest from '../hooks/useRequest';
import AddIcon from '../assets/icons/add.svg';
import DownloadIcon from '../assets/icons/download.svg';

function TestInfo({ examName, completed, order }) {
  const { session } = useSession();
  const { response, loading, error } = useRequest(
    session.token,
    `${process.env.NEXT_PUBLIC_API_URL}/exams?name=${examName}`,
    1,
  );
  if (loading) {
    return <div className="loader" />;
  }

  if (error) {
    return <h3>{error.message}</h3>;
  }

  // PENDING UNTIL THE PART OF DOWNLOAD IS DONE
  async function downloadTest(orderId) {
    const orderBody = { orederIds: orderId };
    try {
      const URL = `${process.env.NEXT_PUBLIC_API_URL}/pdfs`;
      const config = {
        headers: { Authorization: `Bearer ${session.token}` },
      };
      const res = await axios.post(URL, orderBody, config);
      // setTestResponse(res);
    } catch (err) {
      // setTestError(err.data.message);
    }
  }
  // PENDING UNTIL THE ASSIGN PAGE IS READY
  function assignResult(orderId) {
    Router.push(`/add-result/${orderId}`);
  }

  return (
    <section className="test-detail__data">
      <div className="test-detail__data__container--principal">
        <h1 className="test-detail__data__title">Test Information</h1>
        {!completed && session.user.typeOfUser === 'Bacteriologist' ? (
          <AddIcon
            className="test-detail__data__option"
            onClick={assignResult(order)}
          />
        ) : (
          <DownloadIcon
            className="test-detail__data__option"
            onClick={downloadTest(order)}
          />
        )}
      </div>
      {completed && response.data ? (
        <>
          {response.data.map((test) => (
            <>
              <div className="test-detail__data__container">
                <div className="test-detail__data__element">
                  <strong className="test-detail__data__element__head">Test name</strong>
                  <span className="test-detail__data__element__body">{test.name}</span>
                </div>
                <div className="test-detail__data__element">
                  <strong className="test-detail__data__element__head">Test short name</strong>
                  <span className="test-detail__data__element__body">{test.shortName}</span>
                </div>
              </div>
              <div className="test-detail__data__element">
                <strong className="test-detail__data__element__head">Description</strong>
                <span className="test-detail__data__element__body">{test.description}</span>
              </div>
              <div className="test-detail__data__container">
                <div className="test-detail__data__element">
                  <strong className="test-detail__data__element__head">Indications</strong>
                  <span className="test-detail__data__element__body">{test.indications}</span>
                </div>
                <div className="test-detail__data__element">
                  <strong className="test-detail__data__element__head">Schedule minimum days</strong>
                  <span className="test-detail__data__element__body">{`${test.scheduledDays} days`}</span>
                </div>
              </div>
            </>
          ))}
        </>
      ) : (
        ''
      )}
    </section>
  );
}

TestInfo.propTypes = {
  examName: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
  order: PropTypes.string.isRequired,
};

export default TestInfo;
