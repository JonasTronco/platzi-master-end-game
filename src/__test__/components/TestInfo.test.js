import TestInfo from '../../components/TestInfo';
import { create } from 'react-test-renderer';
import Router from 'next/router';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { SessionProvider } from '../../context/SessionContext';

describe('TestInfo Component', () => {
  const mockedRouter = { push: () => {} };
  Router.router = mockedRouter;
  const obj = {
    session: {
      token:
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1ZWNkYTc2NjYxYjE3ZTUwYzg3NzIxYWIiLCJ1c2VybmFtZSI6Im1heW5lLnNuYXNlbC40MTcxIiwic2NvcGVzIjpbInNpZ25pbjphdXRoIiwic2lnbnVwOmF1dGgiLCJyZWFkOm1vdmllcyIsInJlYWQ6dXNlci1tb3ZpZXMiLCJjcmVhdGU6dXNlci1tb3ZpZXMiLCJkZWxldGU6dXNlci1tb3ZpZXMiXSwiaWF0IjoxNTkwNzI1MTIwLCJleHAiOjE1OTExNTcxMjB9.UjjAPEHWbqyPEwX6qcs2UhLvQ6rAeQIL-ME1LDnWBRo',
      user: {
        id: '5ecda76661b17e50c87721ab',
        username: 'mayne.snasel.4171',
        typeOfUser: 'Patient',
        isActive: true,
        imageURL: 'http://dummyimage.com/200x200.png/ff4444/ffffff',
        firstName: 'Mayne',
        lastName: 'Snasel',
        defaultPath: '/tests'
      }
    }
  };

  test('Should render TestInfo component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <TestInfo examName='' order={''} completed={true} />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
