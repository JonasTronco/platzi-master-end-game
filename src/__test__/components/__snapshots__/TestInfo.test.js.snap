// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestInfo Component Should render TestInfo component 1`] = `
<section
  className="test-detail__data"
>
  <div
    className="test-detail__data__container--principal"
  >
    <h1
      className="test-detail__data__title"
    >
      Test Information
    </h1>
    <button
      className="test-detail__data__option--button"
      onClick={[Function]}
      type="button"
    >
      <div
        className="test-detail__data__option"
      />
    </button>
  </div>
  
</section>
`;
